src/lightning/fabric/strategies/model_parallel.py

```diff
@@ -0,0 +1,331 @@
+        _distributed_checkpoint_save(state, path)
+
+    @override
+    def load_checkpoint(
+        self,
+        path: _PATH,
+        state: Optional[Union[Module, Optimizer, Dict[str, Union[Module, Optimizer, Any]]]] = None,
+        strict: bool = True,
+    ) -> Dict[str, Any]:
+        if isinstance(state, (Module, Optimizer)):
+            raise NotImplementedError(
+                "Loading a module or optimizer object from a checkpoint directly is not yet supported."
+            )
+        if strict is False:
+            raise NotImplementedError(f"Non-strict loading is not yet supported in {self.__class__.__name__}.")
+
+        # broadcast the path from rank 0 to ensure all the states are loaded from a common path
+        path = Path(self.broadcast(path))
+        _distributed_checkpoint_load(state, path)  # type: ignore[arg-type]
+        return {}
+
+    def _setup_distributed(self) -> None:
+        reset_seed()
+        self._set_world_ranks()
+        self._process_group_backend = self._get_process_group_backend()
+        assert self.cluster_environment is not None
+        _init_dist_connection(self.cluster_environment, self._process_group_backend, timeout=self._timeout)
+
+    def _setup_device_mesh(self) -> None:
+        from torch.distributed.device_mesh import init_device_mesh
+
+        if self._data_parallel_size == "auto":
+            self._data_parallel_size = self.num_nodes
+        if self._tensor_parallel_size == "auto":
+            self._tensor_parallel_size = self.num_processes
+        if self._data_parallel_size * self._tensor_parallel_size != self.world_size:
+            raise RuntimeError(
+                f"The sizes `data_parallel_size={self._data_parallel_size}` and"
+                f" `tensor_parallel_size={self._tensor_parallel_size}` multiplied should equal the world size"
+                f" ({self.world_size})."
+            )
+        self._device_mesh = init_device_mesh(
+            device_type=self.root_device.type,
+            mesh_shape=(self._data_parallel_size, self._tensor_parallel_size),
+            mesh_dim_names=("data_parallel", "tensor_parallel"),
+        )
+
+    def _get_process_group_backend(self) -> str:
+        return self._process_group_backend or _get_default_process_group_backend_for_device(self.root_device)
+
+    def _set_world_ranks(self) -> None:
+        if self.cluster_environment is not None:
+            self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
+            self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
+        # `LightningEnvironment.set_global_rank` will do this too, but we cannot rely on that implementation detail
+        # additionally, for some implementations, the setter is a no-op, so it's safer to access the getter
+        rank_zero_only.rank = utils_rank_zero_only.rank = self.global_rank
+
+
+def _materialize_module(module: Module, device: torch.device) -> None:
```

Comment:
Couldn't you update and use https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/fabric/utilities/init.py#L59?

Reply:
The logic looks similar but is different. I need to call `to_empty()` on the root (with recurse=True) otherwise it won't work properly with FSDP. The function you reference does it the other way around, it loops over modules and calls `to_empty()` with recurse=False.

Do you want me to replace your implementation with mine?

Reply:
Okay. No need to replace them. I would still move this one together with the others and maybe unify the `reset_parameters` checking logic that are in both

Reply:
Okay, I will move it there. But I also just noticed the function `fabric.utilities.init._materialize_meta_tensors` is not used anywhere in Lightning. Only in LitGPT's sequential.py. Maybe we can revisit this when we have fabric.materialize()